src/components/notebook-utils.js

```diff
@@ -0,0 +1,196 @@
+          style: { marginLeft: '1rem', display: 'flex' },
+          disabled: creating
+        },
+        creating ?
+          [
+            spinner({ size: '1em', style: { color: 'white', marginRight: '1em' } }),
+            'Creating Notebook...'
+          ] :
+          'New Notebook'),
+      Utils.cond(
+        [
+          notebookFailure,
+          () => Modal({
+            title: 'Notebook Creation Failure',
+            okButton: buttonPrimary({ onClick: () => this.setState({ notebookFailure: null }) },
+              'Done'),
+            showCancel: false
+          }, notebookFailure)
+        ],
+        [
+          modalOpen,
+          () => Modal({
+            onDismiss: () => this.setState({ modalOpen: false }),
+            title: 'Create New Notebook',
+            okButton: buttonPrimary({
+              disabled: !(notebookName && notebookKernel),
+              onClick: () => {
+                this.setState({ modalOpen: false, creating: true })
+                Buckets.createNotebook(namespace, bucketName, notebookName, notebookKernel.data,
+                  () => {
+                    this.setState({ creating: false })
+                    reloadList()
+                  },
+                  notebookFailure => this.setState({ notebookFailure, modalOpen: false }))
+              }
+            }, 'Create Notebook')
+          }, [
+            div({ style: Style.elements.sectionHeader }, 'Name'),
+            textInput({
+              style: { margin: '0.5rem 0 1rem' },
+              autoFocus: true,
+              placeholder: 'Enter a name',
+              value: notebookName,
+              onChange: e => this.setState({ notebookName: e.target.value })
+            }),
+            div({ style: Style.elements.sectionHeader }, 'Kernel'),
+            Select({
+              clearable: false,
+              searchable: false,
+              wrapperStyle: { marginTop: '0.5rem' },
+              placeholder: 'Select a kernel',
+              value: notebookKernel,
+              onChange: notebookKernel => this.setState({ notebookKernel }),
+              options: [
+                {
+                  value: 'python2',
+                  label: 'Python 2',
+                  data: python2Notebook
+                },
+                {
+                  value: 'python3',
+                  label: 'Python 3',
+                  data: python3Notebook
+                },
+                {
+                  value: 'r',
+                  label: 'R',
+                  data: rNotebook
+                }
+              ]
+            })
+          ])
+        ],
+        () => null
+      )
+    ])
+  }
+})
+
+export const NotebookDuplicator = hh(class NotebookRenamer extends Component {
+  render() {
+    const { destroyOld, printName, namespace, bucketName, onDismiss, onSuccess } = this.props
+    const { newName, processing, failure } = this.state
+
+    return Modal({
```

Comment:
Discussion: Given the number of parallel conditionals here, this feels like it might want to be two code paths, with the common stuff factored out. (Just an instinct, happy to be convinced otherwise)

Reply:
There are really only 4 places where the split happens, and 3 of them are just string labels. Since the rename and copy methods take the same arguments, I thought it made sense to combine the modals.